File: LeetCode/Hard/2246_Longest_Path_With_Different_Adjacent_Characters/Solution.py
from typing import *

## 2022-04-18 (first attempt using DFS-focused approach with memoization and custom class)
class Node:
    def __init__(self, val):
        self.neighbours = set()
        self.val = val
        self.memo = dict()
    
    def __str__(self):
        return f"val: {self.val}"
class Solution:
    def findLongestPath(self, node, seen):
        if not node:
            return 0
        
        longest = 1
        seen.add(node)
        for neighbour in node.neighbours:
            if neighbour not in seen and neighbour.val != node.val:
                if neighbour in node.memo:
                    longest = max(longest, node.memo[neighbour])
                else:
                    path = 1 + self.findLongestPath(neighbour, seen)
                    node.memo[neighbour] = path
                    longest = max(longest, path)
        return longest
        
    def longestPath(self, parent: List[int], s: str) -> int:
        nodes = [Node(s[i]) for i in range(len(parent))]
        
        for i, p in enumerate(parent):
            if p != -1:
                nodes[i].neighbours.add(nodes[p])
                nodes[p].neighbours.add(nodes[i])
        
        longest = 0
        for node in nodes:
            longest = max(longest, self.findLongestPath(node, set()))
            
        return longest

## 2022-04-18 (better attempt where paths aren't traversed multiple times; better designed Node class)
class Node:
    def __init__(self, val):
        self.children = []
        self.val = val
    
    def __str__(self):
        return f"val: {self.val}"
class Solution:
    longest = 1
    def findLongestPath(self, node):
        if not node:
            return 0
        
        disconnected = [] # longest paths where node.val == child.val
        connected = []    # longest paths where node.val != child.val
        
        for child in node.children:
            if child.val == node.val:
                disconnected.append(self.findLongestPath(child))
            else:
                connected.append(self.findLongestPath(child))
        
        disc = max(disconnected) if disconnected else 1
        conn = 1 + max(connected) if connected else 1
        self.longest = max(max(disc, conn), self.longest)

        # if we have multiple in connected (multiple children where node.val != child.val)
        # then we can connect the two greatest paths into 1 long path
        if len(connected) >= 2:
            connected = sorted(connected)
            self.longest = max(self.longest, connected[-1] + 1 + connected[-2])
        
        return conn
        
    def longestPath(self, parent: List[int], s: str) -> int:
        nodes = [Node(s[i]) for i in range(len(parent))]
        
        for i, p in enumerate(parent):
            if p != -1:
                nodes[p].children.append(nodes[i])
        
        self.findLongestPath(nodes[0])
        return self.longest

## 2022-04-18 (cleaned-up, sexier version of above with some dank list comp and without custom class)
class Solution:
    longest = 1
    def findLongestPath(self, node, parent, children, s):
        disconnected = [self.findLongestPath(child,parent, children, s) \
                        for child in children[node] if s[child] == s[node]]
        connected = [self.findLongestPath(child,parent, children, s) \
                     for child in children[node] if s[child] != s[node]]
        
        disc = max(disconnected) if disconnected else 1
        conn = 1 + max(connected) if connected else 1
        self.longest = max(max(disc, conn), self.longest)
        if len(connected) >= 2:
            connected = sorted(connected)
            self.longest = max(self.longest, connected[-1] + 1 + connected[-2])
        
        return conn
        
    def longestPath(self, parent: List[int], s: str) -> int:
        children = [[] for _ in range(len(parent))]
        
        for i, p in enumerate(parent):
            if p != -1:
                children[p].append(i)
        
        self.findLongestPath(0, parent, children, s) ## start at top to bubble downwards
        return self.longest

File: LeetCode/Hard/2246_Longest_Path_With_Different_Adjacent_Characters/Solution.cpp
// 2023-01-12 (DFS)
class Solution {
public:
    int curLongest = 1;
    int longestPath(vector<int>& parent, string s) {
        /*
            input: a list of node parents parent[i] = i's parent, and a string s
            output: the longest path in the tree where no adjacent nodes have the same char

            notes:
            - seems like something where we run the recurisve funcion on the children,
              and pass up maybe the single longest path that starts at the current node up?
            - will probably need a variable on the side to hold our global answer
            - node 0 is always the root
            - probably generate an adjacency list first to make it easier
            
            approach:
            - create an adjacency list to make iterating easier
            - create a recursive function that both returns the longest path starting at the current
              node + assigns the current best answer to the global variable
                - need parameters for adj list, node, parent (to prevent backtracking)
                - get return values from each child, and merges with total length if child val != cur val
                - return either 1 if node matches both children, or 1 + longest child length if not
            - return our global variable afterwards

            analysis:
            - time: O(n) where n=number of nodes
            - space: O(n) where n=number of nodes
        */
        int n = parent.size();
        vector<vector<int>> children(n);
        for (int i=1; i<n; i++) {
            children[parent[i]].push_back(i);
        }

        findLongest(children, 0, -1, s);
        return curLongest;
    }

    int findLongest(vector<vector<int>>& children, int node, int parent, string& s) {
        int singleLongest = 1;
        for (int child : children[node]) {
            if (child == parent)
                continue;
            int length = findLongest(children, child, node, s);
            // if we can connect our child length to our current node
            if (s[child] != s[node]) {
                // only record our longest connected child; greedily find new curLongest using this
                curLongest = max(curLongest, singleLongest+length);
                singleLongest = max(singleLongest, 1+length);
            }
        } 
        return singleLongest;  
    }
};